### ISSUE TITLE
Update to version 3.4 throw ex :[Right-hand side of 'instanceof' is not an object]

### ISSUE THREAD

**Issue by codesculpture** [1904labs/dom-to-image-more#189]
Duplicating this [issue](https://github.com/1904labs/dom-to-image-more/issues/184) since it is closed, but it should be opened.

This is a interesting issue, since a single change did lot of things from this single [commit](https://github.com/1904labs/dom-to-image-more/commit/0f612a3fdbdc9301a30086e9b56aa8e00fd3ebec#diff-270a29e4e36906febc65fe59a2b7035f5a4f079af8ae57680ee0707332268051L669)

`Object.hasOwn(value, 'getRootNode')` is replaced by `'getRootNode' in value`

But which is semantically not equal, as per mdn, 
##### [Object.hasOwn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn)
The Object.hasOwn() static method returns true if the specified object has the indicated property as its own property. If the **property is inherited**, or does not exist, the method returns false.

#### [`in` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in)
The in operator tests if a string or symbol property is present in an object or its prototype chain. If you want to check for only non-inherited properties, use [Object.hasOwn()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn) instead.

That being said, if `obj` has `prop` and `prop` is a inherited member,

`Object.has(obj, prop)` would return `false`, which is correct
whereas,
`'prop' in obj` would return `true`, which is also correct as per spec.

So what happened here is that `getRootNode` is from [`Node`](https://developer.mozilla.org/en-US/docs/Web/API/Node/getRootNode)
But what we have as `value` is [`HTMLElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement)
which inherits `Node`,
<img width="480" alt="Screenshot 2024-09-12 at 11 54 56 PM" src="https://github.com/user-attachments/assets/7c2ee36c-7cd6-4417-a33c-3f0b372d4432">

so `Object.hasOwn(value, 'getRootNode')` probably returns false, since `getRootNode` is a inherited member from `Node` in `HTMLElement`
but `'getRootNode' in value` would happily returns `true`, since that is correct.
So the logic previously returned `false` but now the same logic returning `true`, which changed the execution flow and would execute this block of code 
https://github.com/1904labs/dom-to-image-more/blob/e7d6ab0535a2477eb1a2b7226105fa1b7c6075a1/src/dom-to-image-more.js#L648C1-L655C10

where `global` would be a empty object, if `dom-to-image-more` was loaded through dynamic import using `webpack`, where `webpack` is importing `dom-to-image-more` by `eval(import("https://mycdn.com/dom-to-image-more.min.js"))` , here if a script is executed through `eval` the `this` object would be a empty object `{}` 
For example, 
`eval(this)` would returns `{}`

 so `global` is `{}` at this point, so we are doing `value instanceof global.ShadowRoot` where `global` is `{}` thus the  rhs of `instanceof` would be undefined. Throwing error.

The error occured for me while using bundler such as `webpack` and dynamically imported `dom-to-image-more` using `import()`.

I believe the fix would be
```
 function getWindow(node) {
            const ownerDocument = node ? node.ownerDocument : undefined;
            return (
                (ownerDocument ? ownerDocument.defaultView : undefined) ||   
                window || 
                global
            );
        }
```

i.e the the existence of `window` would be checked before `global` (just swapping two lines)

@IDisposable

**Comment by IDisposable** [comment]
Thanks for the extensive write-up. Sadly the code you are referencing is not in the current release (neither the `hasOwn`, nor `hasOwnProperty`, nor `in`)

```
        function isInShadowRoot(value) {
            // not calling the method, getting the method
            if (value === null || value === undefined || value.getRootNode === undefined) return false;
            return isShadowRoot(value.getRootNode());
        }
```

While I understand that the referenced commit broke your use-case, we still need to have the shadowed elements code work. Since I _still_ don't have a "this doesn't work" reproduction, I can't actually fix this error.

**Comment by IDisposable** [comment]
I will try swapping those two lines and see if anything breaks.

**Comment by IDisposable** [comment]
There's a prerelease v.3.4.5-beta.0, please try it @codesculpture

**Comment by codesculpture** [comment]
> Thanks for the extensive write-up. Sadly the code you are referencing is not in the current release (neither the `hasOwn`, nor `hasOwnProperty`, nor `in`)
> 
> ```
>         function isInShadowRoot(value) {
>             // not calling the method, getting the method
>             if (value === null || value === undefined || value.getRootNode === undefined) return false;
>             return isShadowRoot(value.getRootNode());
>         }
> ```


This is equivalent to `in`, `in` != `Object.hasOwn`, so this change is also not equivalent to `Object.hasOwn`

Hence, the logic is changed anyway.

**Comment by IDisposable** [comment]
If you can give a PR that adds a test so I can verify the bug and the fix, then I can address the situation. Otherwise, the best I can suggest is using an earlier version. The ShadowDOM fix is not getting backed out for something I cannot reproduce.

Have you tested the 3.4.5-beta.0 release?

**Comment by codesculpture** [comment]
I will try the latest version, and will inform back.
But i think i explained the issue correctly, that it is only happens if you use webpack and dynamically import the library, and webpack is bundling the such as
`const domtoimage = await import('dom-to-image-more')`

to 
`eval('<source>/dom-to-image-more')` and `this` is `Window` Object normally, but inside `eval` it is a empty object `{}`

**Comment by codesculpture** [comment]
Seems 3.4.5-beta.0 fixed the issue, this [commit](https://github.com/1904labs/dom-to-image-more/commit/56a060bc6c90399315d93e899f7a0119db92bd59)

To reproduce this issue,  
create a react app using `npx create-vite`, and choose React, Javascript.
`npm i dom-to-image-more@3.4.4`

And in `App.jsx`
```js
const transform = async (dom) => {
  const domtoimage = await import('dom-to-image-more');
  await domtoimage.toSvg(dom, {});
}
function App() {
  return (
    <div> 
    <div id="hello"> </div>
     <button onClick={() => {
      transform(document.getElementById("hello"))
    }}> Hello </button>
    </div>
  )
}
export default App
```

Click `Hello` button.

##### Expected
No errors
##### Actual 
```
Uncaught (in promise) TypeError: Right-hand side of 'instanceof' is not an object
    at n2 (dom-to-image-more.js:662:20)
    at o2 (dom-to-image-more.js:668:20)
    at Object.isShadowSlotElement (dom-to-image-more.js:709:17)
    at t4 (dom-to-image-more.js:443:26)
    at c2 (dom-to-image-more.js:407:38)
    at dom-to-image-more.js:371:24
    at async transform (App.jsx:3:3)
```

But if you switch to `3.4.5-beta.0` by
`npm i dom-to-image-more@3.4.5-beta.0`

Then if you click `Hello` button,
you no longer see the error.


Seems bundlers (i tested with rollup and webpack), bundles dynamic import by passing custom `this` object which might not be a `window` object, then `global` is what `this` is. So the fix seems okay which gives high priority for `window` over `global` (i.e `window || global` instead of `global || window`)   

@IDisposable Am not sure is it possible to write test case for this scenario, wdyt

**Comment by IDisposable** [comment]
Glad tat it fixed it... we should just hope it broke nothing new... all my existing tests still work.

Released as v3.4.5 (non-beta)

**Comment by codesculpture** [comment]
> Glad that it fixed it... we should just hope it broke nothing new... all my existing tests still work.
> 
> Released as v3.4.5 (non-beta)

If existing tests works, then i guess we are good. Will try to write test that it imports dom-to-image-more using `eval` since thats how bundlers were importing.

Also thanks @IDisposable
